### ISSUE TITLE
Unresolved external symbols related to TLS

### ISSUE THREAD

**Issue by dechenio** [LaurentMazare/tch-rs#563]
I have been struggling to compile the example using prebuilt libtorch+cu117 for Windows (11) with Rust (1.64)-msvc (VS2022). I can build the tch-rs library without error but when linking to any application either my own simple code or the bundled example (i.e. basics), the following errors appeared: 

```shell
$$> $env:TORCH_CUDA_VERSION = "cu117"
$$> $env:LIBTORCH = "c:\dev\libtorch\release"
$$> $env:Path += ";C:\dev\libtorch\release\lib"
$$> $env:RUSTFLAGS = "-L c:\dev\libtorch\release\lib -l c10_cuda"
$$> cargo run --example basics --release
..........
error: linking with `link.exe` failed: exit code: 1120
........
 Creating library D:\projects\rust\tch-rs\target\release\examples\basics.lib and object D:\projects\rust\tch-rs\target\release\examples\basics.exp
          libtorch_sys-18bee47690c056b1.rlib(torch_api.o) : error LNK2001: unresolved external symbol __imp___tls_index_?init@?1??lazy_init_num_threads@internal@at@@YAXXZ@4_NA
          libtorch_sys-18bee47690c056b1.rlib(torch_api.o) : error LNK2001: unresolved external symbol __imp___tls_offset_?init@?1??lazy_init_num_threads@internal@at@@YAXXZ@4_NA
          D:\projects\rust\tch-rs\target\release\examples\basics.exe : fatal error LNK1120: 2 unresolved externals
```

It seems something wrong with linking thread local storage related symbols.  Would you help us to fix it? Many thanks.

**Comment by LaurentMazare** [comment]
I don't know much about windows linking and it seems that it's likely to be an issue with the underlying tls crate rather than with `tch`. That said, afaik `tch` only uses tls through the `curl` dependency which is used to fetch the libtorch binaries automatically. If you have already downloaded libtorch and set the `LIBTORCH` environment variable as seems to be the case you could try disabling this feature. This should hopefully get rid of the `curl` dependency and of tls.
In your `Cargo.toml`, you will need something like:
```
tch = { version = "0.9.0", default-features = false }
```

**Comment by dechenio** [comment]
Thanks @LaurentMazare .
I have tried as above but the same errors related to tls. 
According to what the error indicates, i.e. "libtorch_sys-18bee47690c056b1.rlib(torch_api.o)" . the torch_api.cpp in module torch-sys/libtch may have used some thread local storage features that makes it necessary to have some external symbols (like  __imp___tls_index_ and __imp___tls_offset_) in the linking step in Windows MSVC.

**Comment by LaurentMazare** [comment]
Ah my bad, I assumed that TLS here was there for https connection through curl hence my answer. Thread local storage sounds like a better explanation.
I don't have much to suggest there. Maybe you can try out compiling some basic C++ example using libtorch and see if it works. If that's the case, the flags their cmake config passes to the linker could help figuring out what is missing here (the windows CI tests for the crate also pass properly so something in their setup is likely different from yours).

**Comment by dechenio** [comment]
After doing some investigations, I found this thread in `stackoverflow`: 
[https://stackoverflow.com/questions/74366357/updating-to-visual-studio-17-4-0-yields-linker-errors-related-to-tls](url)
The problem associates with the latest update of MSVC 2022 17.4. VS 17.4 updates STL. 

I have switched to another version like VS 2019 and it works flawlessly. 
Anyway I'm still trying to figure out why the updates of VS 2022 into 17.4 breaks the library.
